Web Inspector: allow protocol domains to be exposed with name that differs from filename

burg opened the pull request:
#### 389769a64d1c48f360ac7807a44be4f3ef37b02e
<pre>
Web Inspector: allow protocol domains to be exposed with name that differs from filename
<a href="https://bugs.webkit.org/show_bug.cgi?id=289657">https://bugs.webkit.org/show_bug.cgi?id=289657</a>

Reviewed by Devin Rousso.

For WebDriver BiDi, we started prefixing the protocol files such as `BidiBrowser.json` with `Bidi` to make these
files easier to find. This prefix was carried over into the &apos;domain&apos; field so as to avoid name conflicts
with Web Inspector&apos;s domains in the generated code, i.e., `BrowserBackendDispatcherHandler` and
`BidiBrowserBackendDispatcherHandler&apos;.

However, we don&apos;t want to expose the commands as `BidiBrowser.close`, rather it should be `browser.close`.
To allow this customization, add a new domain attribute `exposedAs` to control how the domain is referenced externally.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiLog.json:
Properly expose `browser`, `browsingContext`, and `log` domains.

* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py: Register using `exposedAs`.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementations_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_small_dispatcher_switch_implementation_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_large_dispatcher_switch_implementation_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_command):
Update error messages that may be sent externally as protocol errors so that they use `exposedAs` externally.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_implementation.py:
(CppFrontendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_event):
Adopt `domain_exposed_as`.

* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Protocol.parse_domain):
(Domain.__init__):
(Domains):

* Source/JavaScriptCore/inspector/scripts/tests/domain-exposed-as-other-name.json: Added.
* Source/JavaScriptCore/inspector/scripts/tests/expected/domain-exposed-as-other-name.json-result: Added.
Add a test case with a frontend and backend dispatcher to exercise `exposedAs`.

Canonical link: <a href="https://commits.webkit.org/292704@main">https://commits.webkit.org/292704@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7a30c892b892f2c00bb2ac829b14def8f19891d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100068 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45538 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23056 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72493 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29786 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85811 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52820 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10853 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3537 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44877 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87709 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102114 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93660 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22081 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81497 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81829 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80884 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20660 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25447 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2846 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15323 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27191 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/116350 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21711 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33286 "Found 9 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-medium.js.mini-mode, microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/regress/llint-callee-saves-without-fast-memory.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/top-most-enclosing-stack.js.wasm-eager (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25186 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23452 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/scripts/codegen/models.py:
```diff
@@ -429,7 +429,8 @@ def parse_domain(self, json, isSupplemental):
             events.extend([self.parse_event(event, debuggable_types) for event in json['events']])
 
-        self.domains.append(Domain(json['domain'], json.get('description', ''), json.get('condition'), debuggable_types, target_types, isSupplemental, version, types, commands, events))
+        domain_name_exposed_as = json.get('exposedAs', json['domain'])
```
should this be `domainExposedAs`?

lauromoura replied:
Good catch. The test case (`Database` domain), uses `domainExposedAs`, while the BiDi domain json's are using this `exposedAs`.

burg replied:
Ah I see, I wrote the commit message in a hurry. I actually prefer 'exposedAs' to 'domainExposedAs'. How about you?

dcrousso commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
```diff
@@ -1,5 +1,6 @@
 {
     "domain": "BidiBrowser",
+    "exposedAs": "browser",
```
wait why not just have `"domain": "browser"`?

lauromoura replied:
IIUC, the reasoning is due to BiDi having a lot of "generic" modules that might conflict with non-bidi modules. (e.g. we'll have modules like "Network", "Input", etc.)

burg replied:
It is desired to have generated class names to match the filenames because it keeps the Bidi prefix around. Otherwise there is a clash between Inspector's `BrowserBackendDispatcherHandler` and WebDriver's `BrowserBackendDispatcherHandler`.

A lesser reason is that I enjoy having an easy way to grep the generated code for anything WebDriver BiDi related.

AnujPanta1 commented on Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
```diff
@@ -98,6 +98,7 @@ def _generate_dispatcher_implementations_for_domain(self, domain):
 
         constructor_args = {
             'domainName': domain.domain_name,
+            'domainExposedAs': domain.domain_exposed_as
```
I was curious, if we're now using the domainExposedAs variable: how does domainName get used now, and if it's not being updated with the dispatcher how is it initialized ?

lauromoura replied:
`domainName` still gets used as the actual prefix of the generated dispatcher class.

See the collapsed part of `BackendDispatcherImplementationDomainConstructor` template, lines 158-163 of `cpp_generator_templates.py`, above.

lauromoura commented on Source/JavaScriptCore/inspector/scripts/tests/domain-exposed-as-other-name.json:
```diff
@@ -0,0 +1,110 @@
+{
+    "domain": "Database",
+    "domainExposedAs": "legacyDatabase",
```
```suggestion
    "exposedAs": "legacyDatabase",
```

lauromoura commented on Source/JavaScriptCore/inspector/scripts/tests/expected/domain-exposed-as-other-name.json-result:
```diff
@@ -0,0 +1,1758 @@
+        return;
+    }
+
+    m_backendDispatcher->reportProtocolError(BackendDispatcher::MethodNotFound, makeString("'Database."_s, protocol_method, "' was not found"_s));
```
After changing the test json, we need to regenerate this file to get the right "exposed domain name", right?

lauromoura reviewed (commented):
Looks good in general, minor change needed in the test case. (Did not manage to test it yet, though)

dcrousso commented on Source/JavaScriptCore/inspector/scripts/tests/expected/domain-exposed-as-other-name.json-result:
```diff
@@ -0,0 +1,1827 @@
+// DO NOT EDIT THIS FILE. It is automatically generated from domain-exposed-as-other-name.json by the script: Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py
+
+// Database
+InspectorBackend.registerDomain("Database", null);
```
shouldn't all of these be `"legacyDatabase"`?  if the backend is dispatching and receiving using that name then the frontend should also do the same, right?

dcrousso reviewed (approved):
i realize that this is likely only gonna be used by WebDriver BiDi so i think it's fine as-is, but id love to have this be fully implemented for Web Inspector too just in case we ever decide to use it :)

burg committed:
Web Inspector: allow protocol domains to be exposed with name that differs from filename
https://bugs.webkit.org/show_bug.cgi?id=289657

Reviewed by Devin Rousso.

For WebDriver BiDi, we started prefixing the protocol files such as `BidiBrowser.json` with `Bidi` to make these
files easier to find. This prefix was carried over into the 'domain' field so as to avoid name conflicts
with Web Inspector's domains in the generated code, i.e., `BrowserBackendDispatcherHandler` and
`BidiBrowserBackendDispatcherHandler'.

However, we don't want to expose the commands as `BidiBrowser.close`, rather it should be `browser.close`.
To allow this customization, add a new domain attribute `exposedAs` to control how the domain is referenced externally.

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiLog.json:
Properly expose `browser`, `browsingContext`, and `log` domains.

* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py: Register using `exposedAs`.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementations_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_small_dispatcher_switch_implementation_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_large_dispatcher_switch_implementation_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_command):
Update error messages that may be sent externally as protocol errors so that they use `exposedAs` externally.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_frontend_dispatcher_implementation.py:
(CppFrontendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_event):
Adopt `domain_exposed_as`.

* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Protocol.parse_domain):
(Domain.__init__):
(Domains):

* Source/JavaScriptCore/inspector/scripts/tests/domain-exposed-as-other-name.json: Added.
* Source/JavaScriptCore/inspector/scripts/tests/expected/domain-exposed-as-other-name.json-result: Added.
Add a test case with a frontend and backend dispatcher to exercise `exposedAs`.

Canonical link: https://commits.webkit.org/292704@main
